### PR TITLE
Remove references to the structs that have moved to their own packages

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Follows-up https://github.com/kubernetes/kubernetes/pull/66058 and  https://github.com/kubernetes/kubernetes/pull/66059 to remove the structs that now aren't needed in `pkg/apis/componentconfig`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
xref https://github.com/kubernetes/community/pull/2354

**Special notes for your reviewer**:

This PR depends on:
 - [x] https://github.com/kubernetes/kubernetes/pull/67090
 - [x] https://github.com/kubernetes/kubernetes/pull/67149
 - [x] https://github.com/kubernetes/kubernetes/pull/67159
 - [x] https://github.com/kubernetes/kubernetes/pull/67207

**Only review commit 'Remove references to the structs that have moved to their own packages' please**

**Release note**:

```release-note
NONE
```
/kind cleanup
/assign @sttts @thockin @jbeda @liggitt

### CODE REVIEW

**Comment by hanxiaoshuai** [comment]
@luxas It seems some conflicting files needs to be rebase

**Comment by luxas** [comment]
@hanxiaoshuai I'm waiting for https://github.com/kubernetes/kubernetes/pull/66058 and https://github.com/kubernetes/kubernetes/pull/66059 to merge.

/hold
on those PRs merging.

File: pkg/apis/componentconfig/types.go
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: these are the internal types. No need for the json tag.

File: pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go
Comment (sttts) [inline pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go]: was this automatically generated or a manual fixup?

File: pkg/apis/componentconfig/v1alpha1/zz_generated.conversion.go
Comment (sttts) [inline pkg/apis/componentconfig/v1alpha1/zz_generated.conversion.go]: this is probably unexpected. It should call the right conversion func correctly.

**Comment by sttts** [comment]
Some technical issues around generated code.

File: pkg/apis/componentconfig/v1alpha1/zz_generated.conversion.go
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/zz_generated.conversion.go]: likely need to declare a dependency, like https://github.com/kubernetes/kubernetes/pull/66030/files

File: staging/src/k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go]: I didn't expect this change. If we want to make the field consistent, I'd expect us to adjust the go struct field name

File: pkg/apis/componentconfig/types.go
Comment (luxas) [reply pkg/apis/componentconfig/types.go]: oops copy-paste error, fixed

File: pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go
Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go]: automatically generated. As said we might want to revisit this later, but this is as-is for now

File: staging/src/k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go
Comment (luxas) [reply staging/src/k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go]: Fair, I'll do this in a separate PR. The json tag is just wrong, so fixed it up here but will defer that change

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go]: > The json tag is just wrong

I think the json tag is actually what we want, and the field name is wrong

Comment (luxas) [reply staging/src/k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go]: Ok, I don't have any opinions on KubeConfigFile vs KubeConfig, I'll just make it consistent to `KubeConfig` in the next PR then

**Comment by luxas** [comment]
Marking this WIP until the conversion-gen issue @liggitt and @sttts pointed out is resolved

**Comment by luxas** [comment]
> Marking this WIP until the conversion-gen issue @liggitt and @sttts pointed out is resolved

It now is when https://github.com/kubernetes/kubernetes/pull/67159 merges.

**Comment by luxas** [comment]
@sttts @liggitt ready for re-review, the second last commit it is.

File: pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go
Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go]: joining to shared defaults was what @deads2k asked not to do for these common structs in https://github.com/kubernetes/community/pull/2354#issuecomment-407533863

**Comment by liggitt** [comment]
just the comment about the shared defaults on the reusable config structs, lgtm otherwise

**Comment by luxas** [comment]
> just the comment about the shared defaults on the reusable config structs, last two commits lgtm otherwise

@liggitt yes, I know about it and plan to send a PR to fix that. Is that a blocker for this PR to merge, or can it be addressed as a follow up? That requires changes to `k8s.io/apimachinery|apiserver` packages not related to this PR per se.

**Comment by liggitt** [comment]
> That requires changes to k8s.io/apimachinery|apiserver packages not related to this PR per se.

This PR is removing the defaulting functions we'd want to preserve and make effective for the existing structs.

File: pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go
Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/zz_generated.defaults.go]: opened https://github.com/kubernetes/kubernetes/pull/67207 for that

**Comment by luxas** [comment]
@liggitt opened https://github.com/kubernetes/kubernetes/pull/67207 for the defaulting improvements

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/cloud-controller-manager/app/options/options_test.go
Comment (liggitt) [reply cmd/cloud-controller-manager/app/options/options_test.go]: I agree that apiserver is an odd place for leader election config to go (same point made in https://github.com/kubernetes/community/pull/2354#discussion_r202031081).

I can buy that it doesn't really belong under apimachinery either... it's a little higher level than that.

The leader election utilities are under k8s.io/client-go/tools/leaderelection - would it make sense for leader election config to go under client-go somewhere?

**Review by dixudx — APPROVED** [review]
All the dependencies have been merged. It's ready to go.

/lgtm

**Comment by dixudx** [comment]
ping @sttts @deads2k for approval. Thanks.

**Comment by luxas** [comment]
/hold cancel
The PRs that this one depends on have now been merged. This one is ready to go, and will unblock the `kube-scheduler` PR https://github.com/kubernetes/kubernetes/pull/66916 :+1:

**Comment by luxas** [comment]
> This is configuration specific to the controller, naming it like this is super confusing.

@timothysc @liggitt agreed `k8s.io/apiserver` maybe isn't the best possible place for `LeaderElectionConfiguration`, but it's what was initially decided in the KEP. We may still move it before we ship all this, but it's important this PR can merge now, so that dependent PRs get unblocked. The move of `LeaderElectionConfiguration` to e.g. `k8s.io/controller-manager` later can happen in parallel.

> I can buy that it doesn't really belong under apimachinery either... it's a little higher level than that.

Right, hence we initially put this in `k8s.io/apiserver`. `k8s.io/controller-manager` would maybe be even better, however that PR (https://github.com/kubernetes/kubernetes/pull/67015) depends on this one to get merged first.

> The leader election utilities are under k8s.io/client-go/tools/leaderelection - would it make sense for leader election config to go under client-go somewhere?

Maybe, if we think it shouldn't go to `k8s.io/controller-manager` either.

> I think that decision can happen in parallel with this PR, but should be settled before we ship this

Agreed :+1:

**Comment by luxas** [comment]
So TL;DR; I'm asking if this PR can get approved ASAP so it can unblock the rest of the PRs, and then we'll move `LeaderElectionConfiguration` if we want (probably) in parallel.

**Comment by stewart-yu** [comment]
Agree.  Add TODO here, if we find any better file to place `LeaderElectionConfiguration ` at another point in time, we can move it later

File: pkg/apis/componentconfig/v1alpha1/defaults.go
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/defaults.go]: I didn't expect these per-component defaults go away

File: cmd/kube-scheduler/app/options/options_test.go
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options_test.go]: I noticed this test isn't actually inspecting any values of the resulting config, just testing the user credentials.

I expanded it to look at the config that is loaded in https://github.com/kubernetes/kubernetes/pull/67352

That test will help us notice of any defaults change unexpectedly while moving things around, switching to shared structs, etc.

File: pkg/apis/componentconfig/v1alpha1/defaults.go
Comment (stewart-yu) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: Does in Line #274?

Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: That calls RecommendedDefaultClientConnectionConfiguration which might happen to match right now, but I think we want component specific qps/burst defaults. The defaults for the scheduler, kubelet, and controller manager are likely to be different

Comment (stewart-yu) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: Yes, notice it before :+1: 
As comment in https://github.com/kubernetes/kubernetes/pull/66993#pullrequestreview-145902542, we add default value for `kube-controller-manager` and `cloud-controller-manager` in owner `SetDefault_Foo`. I think, we can set `how to fix the issuse` as folllow-up.

Comment (stewart-yu) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: May be we can split the `types.go` and `defaults.go` in different packages, so that we can set specific qps/burst defaults for different component. like that one: https://github.com/kubernetes/kubernetes/tree/master/pkg/apis/batch/v1, `defaults.go` in here, but the `types.go` in `k8s.io/api/batch/v1`.

Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: fixed this by adding the `if zeroValue { default }` statements back for qps and burst which is component-specific.

**Comment by liggitt** [comment]
/assign @bsalamat 
for scheduler cmd package approval

**Review by bsalamat — APPROVED** [review]
/approve

**Pull request merged**
